### ISSUE TITLE
E_WARNING thrown on date validations

### ISSUE THREAD

**Issue by mithredate** [laravel/framework#36138]
- Laravel Version: 8.25.0
- PHP Version: 7.4.4

### Description:
Date validation rules like `after`, `before`, etc. can throw E_WARNING errors if the argument passed in is another data field.

I found other threads related to Laravel 5 with the same problem.

Apparently, the problem is because of [this check](https://github.com/laravel/framework/blob/c26c78cf3bcbaf39e63dcb4a410905c62cb1d282/src/Illuminate/Validation/Concerns/ValidatesAttributes.php#L159-L161).

### Steps To Reproduce:
1- Set `error_reporting` to `E_ALL`
2- Run the following in your `artisan tinker`
```php
Validator::make(['dummy' => Carbon::now()->subDay()->format('Y-m-d H:i:s'), 'another' => Carbon::now()->format('Y-m-d H:i:s')], ['another' => 'after:dummy'])->passes()
```
3- It produces an E_WARNING error like the following:
```
<warning>PHP Warning:  DateTime::modify(): Failed to parse time string (dummy) at position 0 (d): The timezone could not be found in the database in /var/www/vendor/nesbot/carbon/src/Carbon/Traits/Modifiers.php on line 434</warning>
```

Here's a simple snippet that can resolve the issue, but has not been tested thoroughly. I already know that it won't work for `Carbon` instances.
```php
   protected function compareDates($attribute, $value, $parameters, $operator): bool
    {
        if (!is_string($value) && !is_numeric($value) && !$value instanceof DateTimeInterface) {
            return false;
        }

        if ($format = $this->getDateFormat($attribute)) {
            return $this->checkDateTimeOrder($format, $value, $parameters[0], $operator);
        }

        if (
            ($this->hasValue($parameters[0]) &&
                date_create($parameters[0]) === false &&
                date_create($this->getValue($parameters[0]))
            ) || !$date = $this->getDateTimestamp($parameters[0])
        ) {
            $date = $this->getDateTimestamp($this->getValue($parameters[0]));
        }

        return $this->compare($this->getDateTimestamp($value), $date, $operator);
    }

    protected function hasValue($value): bool
    {
        return Arr::has($this->data, $value);
    }
```

**Comment by taylorotwell** [comment]
This doesn't cause any problems in production code where error_reporting is not E_ALL. Laravel disables error reporting in production so nothing is displayed on the screen and we log our own errors.

**Comment by mithredate** [comment]
@taylorotwell There's a [gist](https://gist.github.com/LarsGrevelink/c3e0da3f1642b05ff7a185879a19032a), reporting the same issue in Laravel 5.8. The issue was patched using a solution almost identical to the one I posted above. 

It's not related to the `E_ALL` setting in my dev environment (set the error reporting to E_ALL to make sure I capture E_WARNING errors). We have numerous records of this issue reported in our production logs, i.e. NewRelic, with `error_reporting = E_ALL & ~E_DEPRECATED & ~E_STRICT`.

I can submit a PR and patch this issue.

**Comment by mithredate** [comment]
Here's the link to the issue reported in Laraver 5.8:
https://github.com/laravel/framework/issues/27784

**Comment by LarsGrevelink** [comment]
@mithredate We also saw the issue popping up as soon as we moved to Laravel 8.x. It all seems to originate from PHP's DateTime constructor itself and is logged somewhere under the hood and not catchable even when the reporting value shouldn't allow this. The previously mentioned PHP related issue in #27784 seems to be marked as resolved but something seems to still make its way to the logs anyway.

@taylorotwell This specific occurrence should be fixable by wrapping the usage of the Date facade and mix it with the solution we did for Laravel 5.8 resulting in;

```php
protected function getDateTime($value)
{
    try {
       // Supresses the underlying warning which is entering the log files
        $date = date_create($value);

        if ($date) {
            return Date::parse($date);
        }
    } catch (Exception $e) {
        //
    }
}
```

Or if we want to patch it more centrally wrapping the `DateFactory`'s parse with this same method would do so. Are you open to either of these solutions so that we can fix this?
